src/Symfony/Component/Utf8/GenericStringInterface.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Utf8;
+
+use Symfony\Component\Utf8\Exception\ExceptionInterface;
+
+/**
+ * Represents a generic string implementation.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugo.hamon@sensiolabs.com>
+ *
+ * @throws ExceptionInterface
+ * 
+ * @final
+ */
+interface GenericStringInterface extends \IteratorAggregate
+{
+    public function __toString();
+
+    /**
+     * @return int|null
+     */
+    public function indexOf(string $needle, int $offset = 0);
+
+    /**
+     * @return int|null
+     */
+    public function indexOfIgnoreCase(string $needle, int $offset = 0);
+
+    /**
+     * @return int|null
+     */
+    public function lastIndexOf(string $needle, int $offset = 0);
+
+    /**
+     * @return int|null
+     */
+    public function lastIndexOfIgnoreCase(string $needle, int $offset = 0);
+
+    /**
+     * @return string|null
+     */
+    public function substringOf(string $needle, int $offset = 0);
+
+    /**
+     * @return string|null
+     */
+    public function substringOfIgnoreCase(string $needle, int $offset = 0);
+
+    /**
+     * @return string|null
+     */
+    public function lastSubstringOf(string $needle, int $offset = 0);
+
+    /**
+     * @return string|null
+     */
+    public function lastSubstringOfIgnoreCase(string $needle, int $offset = 0);
+
+    public static function compare(string $a, string $b): int;
+
+    public static function compareNatural(string $a, string $b): int;
+
+    public static function compareIgnoreCase(string $a, string $b): int;
+
+    public static function compareNaturalIgnoreCase(string $a, string $b): int;
+
+    /**
+     * @return static
+     */
+    public function reverse();
+
+    public function length(): int;
+
+    public function width(): int;
```

Comment:
we'll need an arg here, to allow for non-countoing for CLI decorations - could even be the default to me